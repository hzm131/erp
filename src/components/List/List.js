import React, { PureComponent, Fragment } from 'react'
import PropTypes from 'prop-types'
import { Button, Table } from 'antd'
import format from 'date-fns/format'

//const rowSelection = {
//onChange: (selectedRowKeys, selectedRows) => {
//console.log(
//`selectedRowKeys: ${selectedRowKeys}`,
//'selectedRows: ',
//selectedRows
//)
//},
//getCheckboxProps: record => ({
//disabled: false,
//name: record.name
//})
//}

class List extends PureComponent {
  getRows() {
    const { rows } = this.props

    return rows.map(d => {
      d.key = d.id
      return d
    })
  }

  getFields() {
    const { fields } = this.props

    return fields.filter(v => v.fields.list).map(v => {
      return {
        title: v.name,
        dataIndex: v.field,
        render: vv => {
          if (v.type === 'date-time') {
            if (process.env.NODE_ENV === 'test') {
              return '1999-09-09 09:09:09'
            }

            return format(vv, 'YYYY-MM-DD HH:mm:ss')
          }
          return vv
        }
      }
    })
  }

  render() {
    const { pagination, onClickCreate, onClickRow, loading } = this.props

    return (
      <Fragment>
        <Button
          type="primary"
          size="small"
          style={{
            fontSize: 12,
            height: 28,
            width: 44,
            margin: '6px 8px'
          }}
          onClick={onClickCreate}
        >
          创建
        </Button>
        <div>
          <Table
            onRow={record => {
              return {
                onClick: () => onClickRow(record)
              }
            }}
            className="styled-table"
            //rowSelection={rowSelection}
            columns={this.getFields()}
            bordered
            dataSource={this.getRows()}
            size="small"
            pagination={{
              showTotal: (total, range) => `${range[0]}-${range[1]}/${total}`,
              pageSize: 25,
              ...pagination
            }}
            loading={loading}
          />
        </div>
      </Fragment>
    )
  }
}

List.propTypes = {
  /**
   * 数据对象 默认以id为key 所以字段不能以key做field
   */
  rows: PropTypes.arrayOf(PropTypes.object),
  /**
   * 数据字段的定义
   */
  fields: PropTypes.arrayOf(PropTypes.object),
  /**
   * 点新建时调用的回调
   */
  onClickCreate: PropTypes.func,
  /**
   * 点击每行时触发的函数，第一个参数为当前行的记录
   */
  onClickRow: PropTypes.func,
  /**
   * pagigation 的配置项
   */
  pagination: PropTypes.object,
  /**
   * 是否正在加载数据
   */
  loading: PropTypes.bool
}

List.defaultProps = {
  loading: false
}

export default List
