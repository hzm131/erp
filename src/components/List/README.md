List example:

```
const mock = require('./mock').default
;
<List
  rows={mock.rows}
  fields={mock.fields}
  onClickCreate={mock.onClickCreate}
  onClickRow={console.log}
  pagination={{
    pageSize: 3,
    total: 30,
    onChange: console.log
  }}
/>
```

参数示例
```js { "file": "../mock.js" }
```
