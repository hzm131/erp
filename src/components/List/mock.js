import yaml from 'js-yaml'
export default {
  rows: [
    {
      id: 1,
      name: '上海测试科技有限公司',
      abbreviation: '测试',
      established_at: '2018-08-15T08:45:48.961Z',
      created_at: '2018-08-15T08:45:48.961Z',
      hide_value: 'secret'
    },
    {
      id: 2,
      name: 'Apple Inc',
      abbreviation: 'APPL',
      established_at: '2018-08-15T08:45:48.961Z',
      created_at: '2018-08-15T08:45:48.961Z',
      hide_value: 'secret'
    },
    {
      id: 3,
      name: 'Google Inc',
      abbreviation: 'GOOL',
      established_at: '2018-08-15T08:45:48.961Z',
      created_at: '2018-08-15T08:45:48.961Z',
      hide_value: 'secret'
    }
  ],
  fields: yaml.safeLoad(`
---
- name: 机构ID
  field: id
  col:
    span: 12
  rules: []
  required: true
  fields:
    create: true
    edit: true
    detail: true
    list: true
  type: string
- name: 机构名称
  field: name
  col:
    span: 12
  rules:
  - required: true
    message: 必须输入机构名称
  required: true
  fields:
    create: true
    edit: true
    detail: true
    list: true
  type: string
- name: 机构简称
  field: abbreviation
  col:
    span: 12
  rules: []
  required: false
  fields:
    create: true
    edit: false
    detail: true
    list: true
  type: string
- name: 成立时间
  field: established_at
  col:
    span: 12
  rules: []
  required: false
  fields:
    create: true
    edit: true
    detail: true
    list: false
  type: date-time
- name: 创建时间
  field: created_at
  col:
    span: 12
  rules: []
  required: false
  fields:
    create: false
    edit: false
    detail: true
    list: true
  type: date-time
- name: 隐藏字段
  field: hide_value
  col:
    span: 12
  rules: []
  required: false
  fields:
    create: false
    edit: false
    detail: false
    list: false
  type: string
  `),
  onClickCreate() {
    console.log('clicked create')
  },
  onClickEdit() {
    console.log('clicked edit')
  }
}
