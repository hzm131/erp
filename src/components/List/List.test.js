import React from 'react'
import { mount } from 'enzyme'

import List from './List'
import mock from './mock'

it('should render using shallow', () => {
  const mounted = mount(
    <List
      rows={mock.rows}
      fields={mock.fields}
      onClickCreate={mock.onClickCreate}
      onClickRow={console.log}
      pagination={{
        pageSize: 3,
        total: 30,
        onChange: console.log
      }}
    />
  )

  expect(mounted).toMatchSnapshot()
})
