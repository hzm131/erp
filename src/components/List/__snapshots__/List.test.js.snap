// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render using shallow 1`] = `
Array [
  <button
    className="ant-btn ant-btn-primary ant-btn-sm"
    onClick={[Function]}
    style={
        Object {
            "fontSize": 12,
            "height": 28,
            "margin": "6px 8px",
            "width": 44,
          }
    }
    type="button"
>
    <span
        key=".0"
    >
        创 建
    </span>
</button>,
  <div>
    <div
        className="ant-table-wrapper styled-table"
    >
        <div
            className="ant-spin-nested-loading"
            style={null}
        >
            <div
                className="ant-spin-container"
                key="container"
            >
                <div
                    className="ant-table ant-table-small ant-table-bordered ant-table-scroll-position-left"
                    style={Object {}}
                >
                    <div
                        className="ant-table-content"
                    >
                        <div
                            className="ant-table-body"
                            key="bodyTable"
                            onScroll={[Function]}
                            onWheel={[Function]}
                            style={Object {}}
                        >
                            <table
                                className=""
                                key="table"
                                style={Object {}}
                            >
                                <colgroup>
                                    <col
                                        key="id"
                                        style={
                                            Object {
                                                "minWidth": undefined,
                                                "width": undefined,
                                              }
                                        }
                                    />
                                    <col
                                        key="name"
                                        style={
                                            Object {
                                                "minWidth": undefined,
                                                "width": undefined,
                                              }
                                        }
                                    />
                                    <col
                                        key="abbreviation"
                                        style={
                                            Object {
                                                "minWidth": undefined,
                                                "width": undefined,
                                              }
                                        }
                                    />
                                    <col
                                        key="created_at"
                                        style={
                                            Object {
                                                "minWidth": undefined,
                                                "width": undefined,
                                              }
                                        }
                                    />
                                </colgroup>
                                <thead
                                    className="ant-table-thead"
                                >
                                    <tr
                                        style={
                                            Object {
                                                "height": null,
                                              }
                                        }
                                    >
                                        <th
                                            className=""
                                            key="id"
                                        >
                                            <span
                                                key="id"
                                            >
                                                机构ID
                                            </span>
                                        </th>
                                        <th
                                            className=""
                                            key="name"
                                        >
                                            <span
                                                key="name"
                                            >
                                                机构名称
                                            </span>
                                        </th>
                                        <th
                                            className=""
                                            key="abbreviation"
                                        >
                                            <span
                                                key="abbreviation"
                                            >
                                                机构简称
                                            </span>
                                        </th>
                                        <th
                                            className=""
                                            key="created_at"
                                        >
                                            <span
                                                key="created_at"
                                            >
                                                创建时间
                                            </span>
                                        </th>
                                    </tr>
                                </thead>
                                <tbody
                                    className="ant-table-tbody"
                                >
                                    <tr
                                        className="ant-table-row  ant-table-row-level-0"
                                        onClick={[Function]}
                                        onContextMenu={[Function]}
                                        onDoubleClick={[Function]}
                                        onMouseEnter={[Function]}
                                        onMouseLeave={[Function]}
                                        style={
                                            Object {
                                                "height": null,
                                              }
                                        }
                                    >
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            <span
                                                className="ant-table-row-indent indent-level-0"
                                                style={
                                                    Object {
                                                        "paddingLeft": "0px",
                                                      }
                                                }
                                            />
                                            1
                                        </td>
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            上海测试科技有限公司
                                        </td>
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            测试
                                        </td>
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            1999-09-09 09:09:09
                                        </td>
                                    </tr>
                                    <tr
                                        className="ant-table-row  ant-table-row-level-0"
                                        onClick={[Function]}
                                        onContextMenu={[Function]}
                                        onDoubleClick={[Function]}
                                        onMouseEnter={[Function]}
                                        onMouseLeave={[Function]}
                                        style={
                                            Object {
                                                "height": null,
                                              }
                                        }
                                    >
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            <span
                                                className="ant-table-row-indent indent-level-0"
                                                style={
                                                    Object {
                                                        "paddingLeft": "0px",
                                                      }
                                                }
                                            />
                                            2
                                        </td>
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            Apple Inc
                                        </td>
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            APPL
                                        </td>
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            1999-09-09 09:09:09
                                        </td>
                                    </tr>
                                    <tr
                                        className="ant-table-row  ant-table-row-level-0"
                                        onClick={[Function]}
                                        onContextMenu={[Function]}
                                        onDoubleClick={[Function]}
                                        onMouseEnter={[Function]}
                                        onMouseLeave={[Function]}
                                        style={
                                            Object {
                                                "height": null,
                                              }
                                        }
                                    >
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            <span
                                                className="ant-table-row-indent indent-level-0"
                                                style={
                                                    Object {
                                                        "paddingLeft": "0px",
                                                      }
                                                }
                                            />
                                            3
                                        </td>
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            Google Inc
                                        </td>
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            GOOL
                                        </td>
                                        <td
                                            className=""
                                            onClick={[Function]}
                                        >
                                            1999-09-09 09:09:09
                                        </td>
                                    </tr>
                                </tbody>
                            </table>
                        </div>
                    </div>
                </div>
                <ul
                    className="ant-pagination ant-table-pagination mini"
                    style={Object {}}
                    unselectable="unselectable"
                >
                    <li
                        className="ant-pagination-total-text"
                    >
                        1-3/30
                    </li>
                    <li
                        aria-disabled={true}
                        className="ant-pagination-disabled ant-pagination-prev"
                        onClick={[Function]}
                        onKeyPress={[Function]}
                        tabIndex={null}
                        title="Previous Page"
                    >
                        <a
                            className="ant-pagination-item-link"
                        />
                    </li>
                    <li
                        className="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
                        onClick={[Function]}
                        onKeyPress={[Function]}
                        tabIndex="0"
                        title={1}
                    >
                        <a>
                            1
                        </a>
                    </li>
                    <li
                        className="ant-pagination-item ant-pagination-item-2"
                        onClick={[Function]}
                        onKeyPress={[Function]}
                        tabIndex="0"
                        title={2}
                    >
                        <a>
                            2
                        </a>
                    </li>
                    <li
                        className="ant-pagination-item ant-pagination-item-3"
                        onClick={[Function]}
                        onKeyPress={[Function]}
                        tabIndex="0"
                        title={3}
                    >
                        <a>
                            3
                        </a>
                    </li>
                    <li
                        className="ant-pagination-item ant-pagination-item-4"
                        onClick={[Function]}
                        onKeyPress={[Function]}
                        tabIndex="0"
                        title={4}
                    >
                        <a>
                            4
                        </a>
                    </li>
                    <li
                        className="ant-pagination-item ant-pagination-item-5 ant-pagination-item-before-jump-next"
                        onClick={[Function]}
                        onKeyPress={[Function]}
                        tabIndex="0"
                        title={5}
                    >
                        <a>
                            5
                        </a>
                    </li>
                    <li
                        className="ant-pagination-jump-next"
                        key="next"
                        onClick={[Function]}
                        onKeyPress={[Function]}
                        tabIndex="0"
                        title="Next 5 Pages"
                    >
                        <a
                            className="ant-pagination-item-link"
                        />
                    </li>
                    <li
                        className="ant-pagination-item ant-pagination-item-10"
                        onClick={[Function]}
                        onKeyPress={[Function]}
                        tabIndex="0"
                        title={10}
                    >
                        <a>
                            10
                        </a>
                    </li>
                    <li
                        aria-disabled={false}
                        className=" ant-pagination-next"
                        onClick={[Function]}
                        onKeyPress={[Function]}
                        tabIndex={0}
                        title="Next Page"
                    >
                        <a
                            className="ant-pagination-item-link"
                        />
                    </li>
                </ul>
            </div>
        </div>
    </div>
</div>,
]
`;
